aspnetcore/mvc/advanced/custom-model-binding/sample/CustomModelBindingSample/Binders/AuthorEntityBinder.cs

```diff
@@ -0,0 +1,71 @@
+﻿using CustomModelBindingSample.Data;
+using Microsoft.AspNetCore.Mvc.Internal;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace CustomModelBindingSample.Binders
+{
+    #region demo
+    public class AuthorEntityBinder : IModelBinder
+    {
+        private readonly AppDbContext _db;
+        public AuthorEntityBinder(AppDbContext db)
+        {
+            _db = db;
+        }
+
+        public Task BindModelAsync(ModelBindingContext bindingContext)
+        {
+            if (bindingContext == null)
+            {
+                throw new ArgumentNullException(nameof(bindingContext));
+            }
+
+            // specify a default argument name if none is set by ModelBinderAttribute
+            var modelName = bindingContext.BinderModelName;
+            if (string.IsNullOrEmpty(modelName))
+            {
+                modelName = "authorId";
+            }
+
+            // attempt to fetch the value of the argument by name
+            var valueProviderResult =
+                bindingContext.ValueProvider.GetValue(modelName);
+
+            // check if no matching argument exists
+            if (valueProviderResult == ValueProviderResult.None)
+            {
+                return TaskCache.CompletedTask;
+            }
+
+            bindingContext.ModelState.SetModelValue(modelName,
+                valueProviderResult);
+
+            var value = valueProviderResult.FirstValue;
+
+            // check if the argument is null or empty
+            if (string.IsNullOrEmpty(value))
+            {
+                return TaskCache.CompletedTask;
+            }
+
+            int id = 0;
+            if (!int.TryParse(value, out id))
+            {
+                bindingContext.ModelState.TryAddModelError(
+                                        bindingContext.ModelName,
+                                        "Author Id must be an integer.");
```

Comment:
No, discussed with rynowak. That will return a null which can be used to return a 404, if desired, rather than a 500 error. Non-matching id value isn't an error (in this demo - you could of course implement it differently if business requirements dictated it).